lib/model/message.dart

```diff
@@ -1,21 +1,332 @@
+    }
+    final oldState = outboxMessage.state;
+    // See [OutboxMessageState] for valid state transitions.
+    final isStateTransitionValid = newState != oldState && switch (newState) {
+      OutboxMessageState.hidden => false,
+      OutboxMessageState.waiting =>
+        oldState == OutboxMessageState.hidden,
+      OutboxMessageState.waitPeriodExpired =>
+        oldState == OutboxMessageState.waiting,
+      OutboxMessageState.failed =>
+        oldState == OutboxMessageState.hidden
+        || oldState == OutboxMessageState.waiting
+        || oldState == OutboxMessageState.waitPeriodExpired,
+    };
+    if (!isStateTransitionValid) {
+      throw StateError('Unexpected state transition: $oldState -> $newState');
+    }
+
+    outboxMessage._state = newState;
+    for (final view in _messageListViews) {
+      if (oldState == OutboxMessageState.hidden) {
+        view.addOutboxMessage(outboxMessage);
+      } else {
+        view.notifyListenersIfOutboxMessagePresent(localMessageId);
+      }
+    }
+  }
+
+  /// Send a message and create an entry of [OutboxMessage].
+  Future<void> _outboxSendMessage({
+    required MessageDestination destination,
+    required String content,
+    required String? realmEmptyTopicDisplayName,
+  }) async {
+    assert(!_disposed);
+    final localMessageId = _nextLocalMessageId++;
+    assert(!outboxMessages.containsKey(localMessageId));
+
+    final now = ZulipBinding.instance.utcNow().millisecondsSinceEpoch ~/ 1000;
+    _outboxMessages[localMessageId] = switch (destination) {
+      StreamDestination(:final streamId, :final topic) => StreamOutboxMessage(
+        localMessageId: localMessageId,
+        selfUserId: selfUserId,
+        timestamp: now,
+        conversation: StreamConversation(
+          streamId,
+          topic.processLikeServer(
+            // Processing this just once on creating the outbox message
+            // allows an uncommon bug, because either of these values can change.
+            // During the outbox message's life, a topic processed from
+            // "(no topic)" could become stale/wrong when zulipFeatureLevel
+            // changes; a topic processed from "general chat" could become
+            // stale/wrong when realmEmptyTopicDisplayName changes.
+            //
+            // Shrug. The same effect is caused by an unavoidable race:
+            // an admin could change the name of "general chat"
+            // (i.e. the value of realmEmptyTopicDisplayName)
+            // concurrently with the user making the send request,
+            // so that the setting in effect by the time the request arrives
+            // is different from the setting the client last heard about.
+            zulipFeatureLevel: zulipFeatureLevel,
+            realmEmptyTopicDisplayName: realmEmptyTopicDisplayName),
+          displayRecipient: null),
+        content: content),
+      DmDestination(:final userIds) => DmOutboxMessage(
```

Comment:
How about giving `OutboxMessage` a factory constructor that takes an arbitrary `Conversation`, and dispatches between these two subclass constructors? That would allow this `switch` to get pushed down into computing just `conversation`, and the rest of these two constructor calls to get deduplicated. I think that'd make the structure of what this is doing a bit clearer.

The existing `OutboxMessage` constructor could get renamed to `OutboxMessage._` to make room for the nice factory constructor to have the plain `OutboxMessage` name.